pkg/datapath/linux/config/config.go

```diff
@@ -791,6 +791,14 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		}
 	}
 
+	// Write Identity and ClusterID related macros.
+	cDefinesMap["CLUSTER_ID_MAX"] = fmt.Sprintf("%d", option.Config.MaxConnectedClusters)
+	cDefinesMap["CLUSTER_ID_LEN"] = fmt.Sprintf("%d", identity.GetClusterIDLen())
+
+	identityMax := (1 << identity.GetClusterIDShift()) - 1
+	cDefinesMap["IDENTITY_MAX"] = fmt.Sprintf("%d", identityMax)
+	cDefinesMap["IDENTITY_LEN"] = fmt.Sprintf("%d", identity.GetClusterIDShift())
```

Comment:
> pushing the calculation of IDENTITY_LEN into the datapath
I sort of thought this was what I was doing :sweat_smile: The agent does the calculation and the value is passed into the datapath with this define. Do you have something else in mind?
As for hardcoding, I was really trying to avoid that. This was initially designed to allow extension beyond 511, but it was suggested we should artificially limit it to 255/511 at least initially.

Reply:
Taking a closer look at this, we can move all of these except CLUSTER_ID_MAX into helpers in the C code, however we'll still need to keep `GetClusterIDShift()` in numericidentity.go since that is used to get the max/min allocation identity in the agent.
Doing the same calculation in 2 different places to get the same value feels prone to breaking in some unexpected ways. What do you think @ti-mo ?

Reply:
Makes sense, we can commit to treating it as a config value, it makes sense to ensure the agent and datapath agree on the value. The reason I'm pushing back on this so much is that we want to start converting `#define`s in node_config.h and friends to runtime-provided values, meaning `IDENTITY_LEN` cannot appear in any preprocessing directives. If we merge this as-is, we'll need to convert all preprocessing directives later, with the risk of regressions.
So far most of the values in node_config.h are relatively straightforward; addresses, masks and the like, but the way `IDENTITY_LEN` is used in the C feels a little more intricate.

(By the way, play it safe and use unsigned, fixed-length ints for bitwise ops only, see https://github.com/cilium/cilium/pull/28417. Both in Go and C code, e.g. `GetClusterIDShift` should return a u64 or u32)

To make things easier to convert, I'd suggest this in `node_config.h` instead:

```
DECLARE_CONFIG(u64, CLUSTER_ID_SHIFT, "<some description>")
ASSIGN_CONFIG(CLUSTER_ID_SHIFT, <GetClusterIDShift() value>)
```

(note, I haven't created a helper yet for rendering DECLARE_CONFIG and ASSIGN_CONFIG into node_config.h from Go, please add these next to `defineUint32()` and friends)

Then, to reference it in C code in inlined helpers like `get_identity()`:
`foo = CONFIG(CLUSTER_ID_SHIFT)`

You've already designed your helpers in a way that make this change relatively straightforward. :+1: You'll also need to convert e.g. CLUSTER_ID_UPPER_MASK and MARK_MAGIC_CLUSTER_ID_MASK into inlined helpers since these currently depend on IDENTITY_LEN, which is dynamic.

This will allow someone to come in later and simply remove the `ASSIGN_CONFIG` statement to replace it with a runtime-provided value. Using DECLARE_CONFIG makes CLUSTER_ID_SHIFT show up in BTF info in a specific section as well. For context, the plan is to use code generation to parse this section and emit a single config struct for all config values in the ELF, so we can provide these values in a type-safe manner and effectively bring a datapath config plane to Go.

Reply:
Done! I used macros defined in clustermesh.h for the helpers so that I could override things that depend on a config load in the dummy node_config.h and bpf tests. Not sure if there's a better way to get the global config macros to play nice, let me know how that looks